engine/apps/google/tasks.py

```diff
@@ -15,9 +16,69 @@ def sync_out_of_office_calendar_events_for_user(google_oauth2_user_pk: int) -> N
     google_oauth2_user = GoogleOAuth2User.objects.get(pk=google_oauth2_user_pk)
     google_api_client = GoogleCalendarAPIClient(google_oauth2_user.access_token, google_oauth2_user.refresh_token)
 
-    # NOTE: shift swap request generation will be done in https://github.com/grafana/oncall-private/issues/2590
-    # QUESTION: will we need to persist any information about these calendar events in our database?
-    google_api_client.fetch_out_of_office_events()
+    user = google_oauth2_user.user
+    user_id = user.pk
+
+    logger.info(f"Syncing out of office Google Calendar events for user {user_id}")
+
+    user_google_calendar_settings = user.google_calendar_settings
+    specific_oncall_schedules_to_sync = user_google_calendar_settings["specific_oncall_schedules_to_sync"]
+
+    users_schedules = OnCallSchedule.objects.related_to_user(user)
+    if specific_oncall_schedules_to_sync:
+        users_schedules = users_schedules.filter(public_primary_key__in=specific_oncall_schedules_to_sync)
+
+    for out_of_office_event in google_api_client.fetch_out_of_office_events():
+        event_id = out_of_office_event.raw_event["id"]
+        start_time_utc = out_of_office_event.start_time_utc
+        end_time_utc = out_of_office_event.end_time_utc
+
+        logger.info(
+            f"Processing out of office event {event_id} starting at {start_time_utc} and ending at "
+            f"{end_time_utc} for user {user_id}"
+        )
+
+        for schedule in users_schedules:
+            _, _, upcoming_shifts = schedule.shifts_for_user(
+                user,
+                start_time_utc,
+                datetime_end=end_time_utc,
+            )
+
+            if upcoming_shifts:
+                logger.info(
+                    f"Found {len(upcoming_shifts)} upcoming shift(s) for user {user_id} "
+                    f"during the out of office event {event_id}"
+                )
+
+                shift_swap_request_exists = ShiftSwapRequest.objects.filter(
+                    beneficiary=user,
+                    schedule=schedule,
+                    swap_start=start_time_utc,
+                    swap_end=end_time_utc,
+                ).exists()
```

Comment:
I understand we will not remove swap requests if the OoO events are removed from the Google Calendar, right? (probably complex to detect that if we don't keep a tracking relating swap requests to google events). Probably something we should explicitly mention in docs.

Reply:
Correct.

I was thinking about persisting some information about calendar events, or at least storing the calendar event ID that was used to generate the shift swap, but decided to keep it simple for now. There's also the caveat where if the OOO is updated, it might do some funky things (like generate an overlapping second shift swap for you 🤔).

My hope is that this case isn't that common, but if it pops up in the future, we can augment the logic here